File: services/graphicService/GraphicBuilder.java
package services.graphicService;

import java.awt.Color;
import java.awt.Cursor;
import java.awt.Font;
import java.awt.Image;

import javax.swing.Icon;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;

import static javax.swing.JFrame.EXIT_ON_CLOSE;
import javax.swing.JTextField;
import javax.swing.SwingConstants;
import javax.swing.border.Border;

public final class GraphicBuilder {
    
    private static GraphicBuilder service;

    private GraphicBuilder(){
    }
    
    public Icon getIcon(String path, int width, int height){
        ImageIcon iIcono = new ImageIcon(Icon.class.getResource(path));
        Icon icono = new ImageIcon(iIcono.getImage().getScaledInstance(width, height,Image.SCALE_DEFAULT));
        return icono;
    }
    
    public JPanel setPanel(JPanel panel, int x, int y, int ancho, int largo){
        panel.setBounds(x, y, ancho, largo);
        panel.setLayout(null);
        return panel;
    }
    
    public JPanel getPanel(int x, int y, int ancho, int largo){
        JPanel panel = new JPanel();
        setPanel(panel, x, y, ancho, largo);
        return panel;
    }
    
    public JPanel getPanel(int x, int y, int ancho, int largo, Color color){
        JPanel panel = getPanel(x, y, ancho, largo);
        panel.setBackground(color);
        return panel;
    }
    
    public JLabel setLabel(JLabel label, Icon icono, int x, int y){
        label.setSize(icono.getIconWidth(), icono.getIconHeight());
        label.setLocation(x, y);
        label.setIcon(icono);
        return label;
    }
    
    public JLabel getLabel(Icon icono, int x, int y){
        JLabel label = new JLabel();
        setLabel(label, icono, x, y);
        return label;
    }
    
    public JLabel getLabel(Icon icono, int x, int y, Cursor cursor){
        JLabel label = getLabel(icono, x, y);
        label.setCursor(cursor);
        return label;
    }
    
    public JLabel getLabel(String cadena, int x, int y, int ancho, int alto, Color colorFuente, Font fuente){        
        JLabel label= new JLabel(cadena);
        label.setSize(ancho, alto);
        label.setLocation(x, y);
        label.setForeground(colorFuente);
        label.setFont(fuente);
        return label;
    }
    
    public void setFrame(JFrame frame, int ancho, int alto, String titulo){
        frame.setDefaultCloseOperation(EXIT_ON_CLOSE);
        frame.setTitle(titulo);
        frame.setSize(ancho, alto);
        frame.setLayout(null);
        frame.setLocationRelativeTo(null);
        frame.setVisible(true);
    }
    
    public JFrame getFrame(int ancho, int alto, String titulo){
        JFrame frame = new JFrame();
        setFrame(frame, ancho, alto, titulo);
        return frame;
    }
    
    public JButton getButton(int x, int y, Icon icono, Cursor cursor){
        JButton button = new JButton();
        button.setContentAreaFilled(false);
        button.setBorder(null);
        button.setLocation(x, y);
        button.setCursor(cursor);
        button.setFocusable(false);
        if(icono != null){
            button.setIcon(icono);
            button.setSize(icono.getIconWidth(), icono.getIconHeight());
        }
        return button;
    }
    
    public JButton getButton(int x, int y, int ancho, int alto, Cursor cursor, Icon icono, Color colorFondo, boolean esSolido){
        JButton button = getButton(x, y, icono, cursor);
        button.setSize(ancho, alto);
        button.setBackground(colorFondo);
        button.setContentAreaFilled(esSolido);
        return button;
    }
    
    public JButton getButton(String texto, int x, int y, int ancho, int alto, Cursor cursor, Icon icono, Font fuente, Color colorFondo,
            Color colorFuente,Border borde, String direccion, boolean esSolido){
        
        JButton button = getButton(x, y, ancho, alto, cursor, icono, colorFondo, esSolido);
        button.setText(texto);
        button.setFont(fuente);
        button.setForeground(colorFuente);
        button.setBorder(borde);
        button.setContentAreaFilled(esSolido);
        switch(direccion){
            case "l":
                button.setHorizontalAlignment(SwingConstants.LEFT);
                break;
            case "r":
                button.setHorizontalAlignment(SwingConstants.RIGHT);
                break;    
            default:
                break;
        }
        return button;
    }
    
    public JTextField getJTextField(int x, int y, int width, int height, String text, Color foreground, Color background, Color caret,
            String alignment){
        JTextField textField = new JTextField(text);
        textField.setSize(width, height);
        textField.setLocation(x, y);
        textField.setForeground(foreground);
        textField.setBackground(background);
        textField.setCaretColor(caret);
        switch (alignment) {
            case "CENTER":
                textField.setHorizontalAlignment(0);
                break;
            case "RIGTH":
                textField.setHorizontalAlignment(4);
                break;
            case "TOP":
                textField.setHorizontalAlignment(1);
                break;
            case "BOTTOM":
                textField.setHorizontalAlignment(3);
                break;
            default:
                textField.setHorizontalAlignment(2);
                break;
        }
        return textField;
    }
    
    public static GraphicBuilder getService(){
        if(service == null)
            service = new GraphicBuilder();
        return service;
    }

}

File: services/graphicService/GraphicBuilder.kt
package services.graphicService

import java.awt.*
import javax.swing.*
import javax.swing.border.Border

fun getScaledIcon(path: String, width: Int, height: Int): ImageIcon {
    val iIcon = ImageIcon(path)
    return ImageIcon(iIcon.image.getScaledInstance(width, height, Image.SCALE_DEFAULT))
}

fun JPanel.setProperties(x: Int, y: Int, width: Int, height: Int, color: Color?, layoutManager: LayoutManager?, border: Border? = null) {
    this.setBounds(x, y, width, height)
    this.background = color
    this.border = border
    this.layout = layoutManager
}

/**
 * icon label
 */
fun JLabel.setProperties(x: Int, y: Int, icon: Icon, cursor: Cursor? = null) {
    this.setSize(icon.iconWidth, icon.iconHeight)
    this.setLocation(x, y)
    this.icon = icon
    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
    this.cursor = cursor
}

/**
 * text label
 */
fun JLabel.setProperties(x: Int, y: Int, width: Int, height: Int, str: String?, fontColor: Color?, font: Font?, hAlignment: String? = "LEFT") {
    this.setBounds(x, y, width, height)
    this.foreground = fontColor
    this.font = font
    this.text = str
    this.horizontalAlignment = when(hAlignment) {
        "CENTER" -> SwingConstants.CENTER
        "RIGHT" -> SwingConstants.RIGHT
        else -> SwingConstants.LEFT
    }
}

fun JFrame.setProperties(width: Int, height: Int, layoutManager: LayoutManager?, background: Color? = null, title: String? = null) {
    this.setSize(width, height)
    this.setLocationRelativeTo(null)
    this.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
    this.title = title
    this.layout = layout
    this.background = background
    this.isVisible = true
}

/**
 * Icon button
 */
@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
fun JButton.setProperties(x: Int, y: Int, icon: Icon?, cursor: Cursor?) {
    this.setLocation(x, y)
    this.isContentAreaFilled = false
    this.border = null
    this.cursor = cursor
    this.isFocusable = false
    if (icon != null) {
        this.setSize(icon.iconWidth, icon.iconHeight)
        this.icon = icon
    }
}

/**
 * Button filled with icon
 */
fun JButton.setProperties( x: Int,y: Int, width: Int, height: Int, cursor: Cursor?, background: Color?, isSolid: Boolean, icon: Icon? = null) {
    this.setProperties(x, y, icon, cursor)
    this.setSize(width, height)
    this.background = background
    this.isContentAreaFilled = isSolid
}

/**
 * Text button
 */
fun JButton.setProperties(text: String?, x: Int, y: Int, width: Int, height: Int, cursor: Cursor?, font: Font?, background: Color?, foreground: Color?,
          border: Border?, alignment: String?, isSolid: Boolean) {
    setProperties(x, y, width, height, cursor, background, isSolid)
    this.text = text
    this.font = font
    this.foreground = foreground
    this.border = border
    this.isContentAreaFilled = isSolid
    this.horizontalAlignment = when (alignment) {
        "LEFT" -> SwingConstants.LEFT
        "RIGHT" -> SwingConstants.RIGHT
        else -> SwingConstants.CENTER
    }
}

fun JTextField.setProperties(x: Int, y: Int, width: Int, height: Int, text: String? = "", foreground: Color? = BLACK, background: Color? = WHITE,
         caret: Color? = BLACK, border: Border? = null, hAlignment: String? = "LEFT") {
    this.setBounds(x, y, width, height)
    this.text = text
    this.foreground = foreground
    this.background = background
    this.caretColor = caret
    this.border = border
    this.horizontalAlignment = when (hAlignment) {
        "CENTER" -> SwingConstants.CENTER
        "RIGHT" ->  SwingConstants.RIGHT
        else ->  SwingConstants.LEFT
    }

}
